### PR TITLE
Remove duplicate P256Keypair::ECDSA_sign_hash code

### LINKED ISSUE #18430: `ECPKeypair::ECDSA_sign_hash` should be deprecated since not supported by security-critical APIs in Java
#### Problem
- Spec only defines `Crypto_Sign(privateKey, message)` primitive
- CASE and all other areas where signatures are generated according to the spec use the message variant
- The sign_hash operation is a low-level consideration of crypto APIs and is **sometimes not exposed by platform crypto APIs**. For example, in Java's `java.security.Signature`, including on Android, there is no way to get a signature by providing a pre-computed digest. The API expects digesting to be done by the signing operation.

Overall, replacement of this primitive is trivial wherever used, and would reduce reliance on native implementation that prevent going to hardware/OS-aided key storage, especially where Java is concerned.

 #### Proposed Solution

- Mark `ECDSA_sign_hash` as deprecated
- Replace usages of `ECDSA_sign_hash` with `ECDSA_sign_message` where possible

### PR DESCRIPTION
**Pull request by tcarmelveilleux**
#### Problem
- The ECDSA_sign_hash method is a near identical copy of of ECDSA_sign_msg,
  that takes a raw hash.
- This is problematic since some platforms, like Android, cannot directly sign
  a pre-computed hash with OS-aided APIs, and overall this is not consistent
  with signature APIs that work on messages, and where a digest is an internal
  implementation detail.
- Overall, the method adds little value and prevents easy transition to different
  signing algorithms over time if the hash assumption is kept

Fixes #18430

#### Change overview
- Removes the ECDSA_sign_hash API
- Replaces its usage throughout the SDK
- Updates all tests
- Leaves the ECDSA_verify_hash_signature (since it's only used in one place,
  already in native code, and always against raw public keys)

#### Testing
- Cert tests still pass, including device attestation during commissioning
- Unit tests still pass including updated unit tests

### CODE REVIEW

**Review by msandstedt — APPROVED** [review]

File: src/app/clusters/operational-credentials-server/operational-credentials-server.cpp [hunk 1]
```diff
@@ -985,7 +964,7 @@ bool emberAfOperationalCredentialsClusterAttestationRequestCallback(app::Command
 
     attestationElementsLen = TLV::EstimateStructOverhead(certDeclSpan.size(), attestationNonce.size(), sizeof(uint64_t) * 8);
 
-    if (!attestationElements.Alloc(attestationElementsLen))
+    if (!attestationElements.Alloc(attestationElementsLen + Crypto::kAES_CCM128_Key_Length))
```
Comment (bzbarsky-apple) [inline src/app/clusters/operational-credentials-server/operational-credentials-server.cpp:967]:
```suggestion
    if (!attestationElements.Alloc(attestationElementsLen + attestationChallenge.size()))
```
is more clear, I would think....

File: src/app/clusters/operational-credentials-server/operational-credentials-server.cpp [hunk 2]
```diff
@@ -1099,7 +1091,7 @@ bool emberAfOperationalCredentialsClusterCSRRequestCallback(app::CommandHandler
                                                           0u               // no vendor reserved data
         );
 
-        if (!nocsrElements.Alloc(nocsrLengthEstimate))
+        if (!nocsrElements.Alloc(nocsrLengthEstimate + Crypto::kAES_CCM128_Key_Length))
```
Comment (bzbarsky-apple) [inline src/app/clusters/operational-credentials-server/operational-credentials-server.cpp:1094]:
```suggestion
        if (!nocsrElements.Alloc(nocsrLengthEstimate + attestationChallenge.size()))
```

File: src/crypto/CHIPCryptoPALmbedTLS.cpp [hunk 3]
```diff
@@ -553,33 +553,20 @@ static inline const mbedtls_ecp_keypair * to_const_keypair(const P256KeypairCont
...
 
-    return ECDSA_sign_hash(&digest[0], sizeof(digest), out_signature);
-#else
-    return CHIP_ERROR_NOT_IMPLEMENTED;
-#endif
-}
-
-CHIP_ERROR P256Keypair::ECDSA_sign_hash(const uint8_t * hash, const size_t hash_length, P256ECDSASignature & out_signature) const
-{
 #if defined(MBEDTLS_USE_TINYCRYPT)
```
Comment (bzbarsky-apple) [inline src/crypto/CHIPCryptoPALmbedTLS.cpp:563]: So if both MBEDTLS_USE_TINYCRYPT and MBEDTLS_ECDSA_C are not defined, what happens?

**Review by bzbarsky-apple — APPROVED** [review]
There are still Darwin and Android bits that do sign_hash that are dead code (and will likely not compile, because they are `override`).

File: src/crypto/CHIPCryptoPALmbedTLS.cpp (hunk 3 above)
Comment (tcarmelveilleux) [reply src/crypto/CHIPCryptoPALmbedTLS.cpp:563]: Fixed

**Pull request merged**